README.md
# Egendata operator

![License](https://flat.badgen.net/github/license/egendata/operator)
![Dependabot](https://flat.badgen.net/dependabot/egendata/operator?icon=dependabot)
![Travis CI](https://flat.badgen.net/travis/egendata/operator?icon=travis)
![Github release](https://flat.badgen.net/github/release/egendata/operator?icon=github)

Backend for managing consents and data flow

## What data is stored and where?

- Personal data is stored encrypted in the `PDS` (eg. Dropbox) specified by the user
- Metadata (consents, permissions, id:s) and client data are stored in `Postgres` (this will probably change)
- Temporary consent requests are stored in `Redis`
- Requests and errors are logged in `elasticsearch/APM` for debugging purposes in test/dev

## Configuration

The application runs with default cofiguration if `NODE_ENV` is `development`
(default) or `test` (default when running tests). When running in `production`
these environment variables *must* be set or the application will not start.

These variables may optionally be overridden in an `.env` file on your dev
machine.

```
# Run mode: development|test|production
NODE_ENV=development

# Host port, defaults to 3000
PORT=3000
# Full host address, defaults to http://[your ip]:[port]
HOST=https://myservice.work

# Postgres
PGHOST='postgres-service'
PGPORT='5432'
PGUSER='some-username'
PGPASSWORD='some-very-secret-password'
PGDATABASE='egendata'

# Application performance management. turned off if val is ''
APM_SERVER=http://apm-service:8200
APM_TOKEN=mysecrettoken

# Operator key (generate your own!!!)
PRIVATE_KEY="-----BEGIN RSA PRIVATE KEY-----\nMIICXQIBAAKBgQDgR0M9RBFT26AseaNYy9g5YX+YQhbskFTAp0aCO6h/8xPAOzkd\n7sO+Lqrx4Ljs8iuUTv7JUbx8+ml+7IMKYjpFx4eFQ1kcEW9IL5xSKKcHt2O359cy\n5KbBExqj4Fr2JkmCE+XcMJa5GGOnpNKB92pmJtfOkjVBVQ280n/j8suyCwIDAQAB\nAoGBAMwGqBl86ZJy0nSDN2EZF5ujoXJ+dOJBrogP5CmnYfL7y3Ttq1kakwFY7PPb\nLf+HkrN5ZXj5HVJIb14ihFcW4tBR2EtABhuv2J6ZNx0KnDxUj+mJlb7GNgr5eayI\nUibIu8/eQh2+CGMilI/KR8zlRiHpD8BgttfBaRktGIrzklQJAkEA9C8JgnAGUbPp\n3rc3dEZR6pEcOGI5Fjo3uvhbOYO5oa4tJszNF1Fh1oUmn17J6yoMnh0qPG4snL2B\nOgSB8OCOnwJBAOshovf7obbVZFzQ7ikYImT/pqz7f7eV1+Uv1MRfGsXAc0EAXDrh\nAPiJ5icWkeRDCFxaTAy/8lrDGgDcL2CSoRUCQQCem4L4x91C6rMJaEbL7vU8gL8s\n3JgqGOykNLfElwxXubQ4VKUO9Vywo9JfiIlth+WkOlt53zJ5KRqsXcstdB8PAkAo\nw6IfYA6/Reuqc8Z2dWqxG+lnoAqaZ24Qm+RFTz+y/RR+NnPG+W9Tp4SxTiZo7n4q\nlLUOmNCJj72YXJQSKBmpAkAyDc4PrJ3nFt45BOEnRuXE60Lv3VzLPdWggOLcKTbW\nr6NAWQS0VNdXEmJVmdoKFhJAeUvLrXPtBGqPS7HO6A8A\n-----END RSA PRIVATE KEY-----\n"
```

## Logging

The Operator is instrumented to use [Elastic APM](https://www.elastic.co/products/apm)
for performance and error logging.


CHANGELOG.md
## [0.33.1](https://github.com/egendata/operator/compare/v0.33.0...v0.33.1) (2020-01-15)


### Bug Fixes

* 🐛 Updated insecure dependendies ([#105](https://github.com/egendata/operator/issues/105)) ([0a93a0e](https://github.com/egendata/operator/commit/0a93a0ebaeb0c5090a6f3a741b5e527f01a6e6ab))

# [0.33.0](https://github.com/egendata/operator/compare/v0.32.1...v0.33.0) (2020-01-15)


### Features

* 🎸 read recipients ([#94](https://github.com/egendata/operator/issues/94)) ([42b87cc](https://github.com/egendata/operator/commit/42b87cc5477e24c02893dcb481379ea5233f525e))

## [0.32.1](https://github.com/egendata/operator/compare/v0.32.0...v0.32.1) (2019-10-30)


### Bug Fixes

* 🐛 revert memory-fs bump ([#74](https://github.com/egendata/operator/issues/74)) ([c2dfffb](https://github.com/egendata/operator/commit/c2dfffba8017459daf6a5b8d30b9911d1c204950))

# [0.32.0](https://github.com/egendata/operator/compare/v0.31.1...v0.32.0) (2019-10-24)


### Features

* 🎸 remove word api from routes ([#58](https://github.com/egendata/operator/issues/58)) ([33d77e3](https://github.com/egendata/operator/commit/33d77e329076930ead22b7f6770a824409085c0c))

## [0.31.1](https://github.com/egendata/operator/compare/v0.31.0...v0.31.1) (2019-10-09)


### Bug Fixes

* 🐛 setting the keys correctly for production NODE_ENV ([#52](https://github.com/egendata/operator/issues/52)) ([cb22cb3](https://github.com/egendata/operator/commit/cb22cb3))

# [0.31.0](https://github.com/egendata/operator/compare/v0.30.4...v0.31.0) (2019-10-02)


### Features

* 🎸 No config needed for development ([#42](https://github.com/egendata/operator/issues/42)) ([8def0b7](https://github.com/egendata/operator/commit/8def0b7))

## [0.30.4](https://github.com/egendata/operator/compare/v0.30.3...v0.30.4) (2019-09-19)


### Bug Fixes

* 🐛 Removed unused redis dependency ([#35](https://github.com/egendata/operator/issues/35)) ([b707454](https://github.com/egendata/operator/commit/b707454))


lib/config.js
require('dotenv').config()
const { JWK } = require('@panva/jose')
const config = require('nconf')
  .use('memory')
  .argv()
  .env()
  .defaults({
    NODE_ENV: 'development',
    APP_NAME: 'egendata-operator'
  })

function exitWith (message) {
  process.stderr.write(`${message}\n`)
  process.exit(1)
}

function privateKey (host, rsaKey) {
  const rxPEM = /^-----BEGIN RSA PRIVATE KEY-----\n([a-zA-Z0-9+/=]*\n)*-----END RSA PRIVATE KEY-----\n?$/
  if (typeof rsaKey !== 'string' || !rxPEM.test(rsaKey)) {
    exitWith('Unknown key format for PRIVATE_KEY')
  }

  const keyConfig = {
    kid: `${host}/jwks/operator_key`,
    use: 'sig'
  }
  return JWK.asKey(rsaKey, keyConfig).toJWK(true)
}

function publicKey ({ e, kid, kty, n, use }) {
  return { e, kid, kty, n, use }
}

function checkMissingKeys (config, keys) {
  const missing = keys.filter(key => {
    return !config.get(key)
  })
  if (missing.length > 0) {
    exitWith(`Your config is missing the following keys: ${missing.join(', ')}`)
  }
}

let ip, PORT, HOST, PGHOST, PGPORT, PGUSER, PGPASSWORD, PGDATABASE, PRIVATE_KEY, PUBLIC_KEY

switch (config.get('NODE_ENV')) {
  case 'development':
    ip = require('ip').address()
    PORT = config.get('PORT') || '3000'
    HOST = `http://${ip}:${PORT}`
    PGHOST = config.get('PGHOST') || ip
    PGPORT = config.get('PGPORT') || '5432'
    PGUSER = config.get('PGUSER') || 'postgresuser'
    PGPASSWORD = config.get('PGPASSWORD') || 'postgrespassword'
    PGDATABASE = config.get('PGDATABASE') || 'egendata'
    PRIVATE_KEY = privateKey(HOST, config.get('PRIVATE_KEY') || '-----BEGIN RSA PRIVATE KEY-----\nMIICXQIBAAKBgQDgR0M9RBFT26AseaNYy9g5YX+YQhbskFTAp0aCO6h/8xPAOzkd\n7sO+Lqrx4Ljs8iuUTv7JUbx8+ml+7IMKYjpFx4eFQ1kcEW9IL5xSKKcHt2O359cy\n5KbBExqj4Fr2JkmCE+XcMJa5GGOnpNKB92pmJtfOkjVBVQ280n/j8suyCwIDAQAB\nAoGBAMwGqBl86ZJy0nSDN2EZF5ujoXJ+dOJBrogP5CmnYfL7y3Ttq1kakwFY7PPb\nLf+HkrN5ZXj5HVJIb14ihFcW4tBR2EtABhuv2J6ZNx0KnDxUj+mJlb7GNgr5eayI\nUibIu8/eQh2+CGMilI/KR8zlRiHpD8BgttfBaRktGIrzklQJAkEA9C8JgnAGUbPp\n3rc3dEZR6pEcOGI5Fjo3uvhbOYO5oa4tJszNF1Fh1oUmn17J6yoMnh0qPG4snL2B\nOgSB8OCOnwJBAOshovf7obbVZFzQ7ikYImT/pqz7f7eV1+Uv1MRfGsXAc0EAXDrh\nAPiJ5icWkeRDCFxaTAy/8lrDGgDcL2CSoRUCQQCem4L4x91C6rMJaEbL7vU8gL8s\n3JgqGOykNLfElwxXubQ4VKUO9Vywo9JfiIlth+WkOlt53zJ5KRqsXcstdB8PAkAo\nw6IfYA6/Reuqc8Z2dWqxG+lnoAqaZ24Qm+RFTz+y/RR+NnPG+W9Tp4SxTiZo7n4q\nlLUOmNCJj72YXJQSKBmpAkAyDc4PrJ3nFt45BOEnRuXE60Lv3VzLPdWggOLcKTbW\nr6NAWQS0VNdXEmJVmdoKFhJAeUvLrXPtBGqPS7HO6A8A\n-----END RSA PRIVATE KEY-----\n')
    PUBLIC_KEY = publicKey(PRIVATE_KEY)

    module.exports = config.defaults({
      PORT,
      HOST,
      PGHOST,
      PGPORT,
      PGUSER,
      PGPASSWORD,
      PGDATABASE,
      PRIVATE_KEY,
      PUBLIC_KEY,
      APM_SERVER: `http://${ip}:8200`,
      APM_TOKEN: 'abc'
    })
    break
  case 'test':
    PORT = '3000'
    HOST = `http://localhost:${PORT}`
    PGHOST = 'localhost'
    PGPORT = '5432'
    PGUSER = 'postgresuser'
    PGPASSWORD = 'postgrespassword'
    PGDATABASE = 'egendata'
    PRIVATE_KEY = privateKey(HOST, '-----BEGIN RSA PRIVATE KEY-----\nMIICXQIBAAKBgQDgR0M9RBFT26AseaNYy9g5YX+YQhbskFTAp0aCO6h/8xPAOzkd\n7sO+Lqrx4Ljs8iuUTv7JUbx8+ml+7IMKYjpFx4eFQ1kcEW9IL5xSKKcHt2O359cy\n5KbBExqj4Fr2JkmCE+XcMJa5GGOnpNKB92pmJtfOkjVBVQ280n/j8suyCwIDAQAB\nAoGBAMwGqBl86ZJy0nSDN2EZF5ujoXJ+dOJBrogP5CmnYfL7y3Ttq1kakwFY7PPb\nLf+HkrN5ZXj5HVJIb14ihFcW4tBR2EtABhuv2J6ZNx0KnDxUj+mJlb7GNgr5eayI\nUibIu8/eQh2+CGMilI/KR8zlRiHpD8BgttfBaRktGIrzklQJAkEA9C8JgnAGUbPp\n3rc3dEZR6pEcOGI5Fjo3uvhbOYO5oa4tJszNF1Fh1oUmn17J6yoMnh0qPG4snL2B\nOgSB8OCOnwJBAOshovf7obbVZFzQ7ikYImT/pqz7f7eV1+Uv1MRfGsXAc0EAXDrh\nAPiJ5icWkeRDCFxaTAy/8lrDGgDcL2CSoRUCQQCem4L4x91C6rMJaEbL7vU8gL8s\n3JgqGOykNLfElwxXubQ4VKUO9Vywo9JfiIlth+WkOlt53zJ5KRqsXcstdB8PAkAo\nw6IfYA6/Reuqc8Z2dWqxG+lnoAqaZ24Qm+RFTz+y/RR+NnPG+W9Tp4SxTiZo7n4q\nlLUOmNCJj72YXJQSKBmpAkAyDc4PrJ3nFt45BOEnRuXE60Lv3VzLPdWggOLcKTbW\nr6NAWQS0VNdXEmJVmdoKFhJAeUvLrXPtBGqPS7HO6A8A\n-----END RSA PRIVATE KEY-----\n')
    PUBLIC_KEY = publicKey(PRIVATE_KEY)

    module.exports = config.defaults({
      PORT,
      HOST,
      PGHOST,
      PGPORT,
      PGUSER,
      PGPASSWORD,
      PGDATABASE,
      PRIVATE_KEY,
      PUBLIC_KEY,
      APM_SERVER: '',
      APM_TOKEN: ''
    })
    break
  default:
    checkMissingKeys(config, ['PORT', 'HOST', 'DATABASE_URL', 'PRIVATE_KEY'])

    config.set('PRIVATE_KEY', privateKey(config.get('HOST'), config.get('PRIVATE_KEY')))
    config.set('PUBLIC_KEY', publicKey(config.get('PRIVATE_KEY')))

    module.exports = config
    break
}
